integration with fabm

plazzari opened the pull request:
close #17

plazzari committed:
Adding preliminary example to couple with FABM

plazzari commented:
The corresponding ModelBuild is https://github.com/inogs/ModelBuild/blob/fabm/downloader_ogstm_bfm.sh

Marco-Puglia committed:
Add leonardo.intel

Add leonardo.intel file to make this branch work on Leonardo server.

Commit:
adding include and library paths for FABM

Commit:
first test including fabm module

Commit:
start including initialization phase of fabm in ogstm

Commit:
Adding Initialize and fabm.yaml read procedure

Commit:
bug fix

Commit:
Bug fix, put linking of fabm library in the correct order, to fulfill dependencies

Commit:
add configuration file fabm.yaml to the FABM

Commit:
change name of lidata add suffix _ogstm to avoid conflict with fabm

Commit:
correct string for FABM path

Commit:
add the range of all bio-optical parameters

Commit:
include separate subroutines for native bfm and fabm

Commit:
select coupling with fabm

Commit:
added more initialization functions, the creation of the model is currently in a bad position with the BIOmem allocation function

Commit:
include switch between bfm old style and fabm

Commit:
remove bfm code when bfm not active

Commit:
Make ogstm agnostic with respect to biogeochemical model used, remove reference to BFM if key_trc_bfm not active, default model is one passive tracer

Commit:
finalize previuos commit: Make ogstm agnostic with respect to biogeochemical model used, remove reference to BFM if key_trc_bfm not active, default model is one passive tracer

Commit:
add test namelists for different BGC-TYPE

Commit:
add configuration column for BGC-TYPE

Commit:
Allows the model to use non parameter definition of BGC matrix sizes

Commit:
Retrive size of passive tracers matrixes from FABM

Commit:
Finalize to pass FABM internal variables names and units to OGSTM, namelist.passivetrc not used any more, diagnostic writing activations and frequency hard coded, need to be solved.

Commit:
set no BGC as Default configuration

plazzari commented:
@gbolzon  For FABM developments, it no longer makes much sense to keep namelist.passivetrc in its current form.

&NATTRC
ctrcnm(1)="N1p"
ctrcun(1)="mmol P/m3"
ctrmax(1)=1.000000e+0
ctr_hf(1)=1
/
Many things are now assigned directly by FABM; for example, ctrcnm and ctrcun are no longer needed.

It is still necessary to specify the assumed maximum value of a state variable ctrmax, whether it should be saved, and at what frequency a given variable should be stored.

Backward compatibility will be ensured when using BFM old, for which the legacy namelist.passivetrc remains valid.

For the new variable declaration, I would propose something like the following:

AVE_FREQ_1
N1_p, N3_n, ...
AVE_FREQ_2
N1_p .....

gbolzon commented:
Actually the info about High-Frequency and max is here
https://github.com/inogs/ogstm/blob/a242f914ae03d1cd8410530727872d1bc4dc229f/bfmv5/BFMtab.xml#L1-L8

At the moment is read in [builder_ogstm_bfm.sh](https://github.com/inogs/ModelBuild/blob/ed90cb691b3cfa528e9b406b5eccdc21e8ff82ec/builder_ogstm_bfm.sh#L198-L204) to create a `namelist.passivetrc` suitable for ogstm

This info could be read by an xml file like this one, maybe directly by ogstm

Commit:
Fix BFM and DEFAULT compilation in parini.f90

plazzari commented:
Comment by Jorn, that suggests to use yaml in place of xml:
You could potentially use FABM’s built-in yaml parser for that purpose, as for instance FVCOM has done:

https://github.com/UK-FVCOM-Usergroup/uk-fvcom/blob/FVCOM-FABM/FVCOM_source/mod_fabm_3D.F#L2763-L2794

 Or for a more comprehensive example of that:

https://github.com/gotm-model/code/blob/master/src/gotm/gotm.F90#L202-L245

Commit:
copy all the namelist need .optics and .phys when using fabm BGC_TYPE

Commit:
correct not ifdef with propoer syntax

Commit:
allow for greater lenght of restart  filenames and variables specification them.

Commit:
include variable naming of FABM-BFM

Commit:
hardcode relaxation variables --> need to be implemented better to adapt to FABM,extend files and variables naming to allow for larger strings

Commit:
adding module to read ogstm.yaml file, to configure I/O activation and frequency.

Commit:
include cmake options to include yaml modules

Commit:
connect configuration parameter in ogstm.yaml for I/O to operating ones in ogstm, relaxation activation needs to be completed

Commit:
add yaml directives to CMAKE

Commit:
set to default no BFM no FABM

plazzari commented:
Current version of the new ogstm.yaml
[login01 OGSTM-FABM]$ cat `ModelBuild/ogstm/testcase/TEST01/wrkdir/MODEL/ogstm.yaml`
```
interior_state:
    O2_o:
        ctrmax: 1.000000e+03 
        ctrhf:  1
        relax:  1
    N1_p:
        ctrmax: 5.000000e+00
        ctrhf:  1
        relax:  1
interior_diagnostic:
    ETW:
        diahf: 0
        diaWR: 0
    ESW:
        diahf: 0
        diaWR: 0

horizontal_diagnostic:
    COSED:
            diahf_2d: 0
            diaWR_2d: 0
    EPCO2air:
            diahf_2d: 0
            diaWR_2d: 0
```

question for @gbolzon,
behaviour of the reader, if a variable is present in the yaml but not in the model
  - skip
  - raise error
 ?

gbolzon commented:
Hi @plazzari, in my opinion 
if a variable is present in the yaml but not in the model, a Warning should be raised

It could be useful to have many yaml files, one for kind of simulation.

Commit:
exclude use of yaml module in default and BFM configurations

Commit:
add controls on ogstm.yaml declaration of tracer, need to be consistent with internal fabm variables

Commit:
script to create template of ogstm.yaml

Commit:
fix path and deployment  for ogstm.yaml

plazzari commented:
@gbolzon, it seems that  variables to be relaxed are currently managed in BC namelist files, so no need to put a configuration parameter in the ogstm.yaml

Can you confirm?

Commit:
define seconds_per_time_unit to prevent error in FABM initialization

Commit:
Dump info about restart files read.

Commit:
Start including the environmental dependencies required by FABM.

Commit:
set always at least one HF output otherwise create problems all outputs at 10.e-08.

Commit:
Bug fix on double declared variable.

Commit:
Finalize the definition of FABM dependencies.

plazzari committed:
Improve ogstm.yaml handling in deploy_code.py

Add error handling for ogstm.yaml copying process and create template if not found.

Commit:
Include instructions to apply derivatives from FABM

Commit:
adding mask for terrain points

Commit:
initialized aksi short wave radiation, huge values gives overflow need to check more in detail

Commit:
initialized diagnostic to zero to prevent overflow, need to put correct values

Commit:
adding diagnostic data from FABM to OGSTM

Commit:
Adding sinking velocity taking the velocity from FABM, all variables possibly sink according to FABM specs

Commit:
adding time variable, need to be fixed

Commit:
passing time parameters to FABM

Commit:
added new function for day of the year

Commit:
pass to FABM the correct time informations

Commit:
from bfm to bgc flag activation

Commit:
add target and pointer definitions

Commit:
setting diagnostic write default to -> false

Commit:
add momentary write

Commit:
graphic interface to configure ogstm.yaml

Commit:
Adding trcave alternative subroutine for FABM

Commit:
deactivaded recovery of diagnostics from restart in FABM

Commit:
added improved diagnostic allocation according to what is dumped by user, correction of linking functions in FABM initialization by passing pure memeory address

Commit:
ogstm.yaml creation with graphic interface

Commit:
added a missing space in the cp yaml instruction

Commit:
switched from the now absent lbfm flag to the new lbgc flag in all the namelist files

Commit:
revert to previous version because the change was useless

Commit:
changed #ifndef to #ifdef to fix the bfm 'zero matrix output' runs

Commit:
correct unit of measure and direction for sinking velocities, FABM provides sinking velocities already in m/s and negative

Commit:
Add RosenMacartur model setup

Commit:
Add case for RosenMacartur mode

Commit:
Add GUI for ogstm.yaml  file configuration

Commit:
extension of variable name strings for output files to 200 digits, bug fix of diadump when high freq not active random output values.

Commit:
bug fix for function DAY_OF_THE_YEAR

Commit:
added different options for var_nc definition of Restoration Mask, this part will be removed, see issue in GitHub

Commit:
Removed write for debug phase

Commit:
set aerosol_air_mass_type from 10 - continental - to 1 - open ocean -

Commit:
Fixed error message to include the new FABM-BFM flag.

Commit:
Select activation of ogstm OASIM model according to FABM.yaml file

Commit:
adding automatic selction of atmospheric model from fabm version or BIOPTIMOD version

plazzari committed:
Create init files for BFM98

plazzari committed:
script to generate vertical profiles for BFM98

Commit:
Added the mapping dictionary creator

plazzari committed:
code cleaning

plazzari committed:
code cleaning

Commit:
Added bgc_error.py to house shared error logic and edited the relative files

Commit:
Bug fix on buondary conditions index i->j

plazzari committed:
update python script with BFM98

plazzari committed:
correct error message

Commit:
Merge remote-tracking branch 'refs/remotes/origin/17-integration-with-fabm' into 17-integration-with-fabm

Commit:
new create_RESTARTS98.py file

Commit:
utility to create boundary conditions namelists with ogstm-fabm

Commit:
refine utility to create boundary conditions namelists with ogstm-fabm

Commit:
new create_ATL98.py file

Marco-Puglia committed:
Update create_ATL98.py

Commit:
Bug Fix for Parallel I/O when CPU per node > 99, more than two digit required.

Commit:
Script to create 1.aveTimes type files

Commit:
Add Sea-Atmosphere exchange fluxes

Commit:
Cleaning

Commit:
added bottom_flux to fabm

Commit:
included WRITE error messages in the key_trc_fabm case

Commit:
included sedimentation speed also for key_trc_fabm after turning on bottom_flux

Commit:
add in-situ water formulation

Commit:
update wind read format

Commit:
minor update on bottom flux parameter initialization

Commit:
vertical velocities definid in namelist.init only for BFM

Commit:
vsed in namelist.init only for BFM

Commit:
align with neccton

Commit:
add in-situ temperature functions

Commit:
bug fix

Commit:
set fabm as default

Commit:
Added wsp10 variable calculated from wind components

Commit:
Better formatting
